csss-site/src/tools/lib/generate_document.py

```diff
@@ -0,0 +1,222 @@
+    #       (or something like that)
+
+    relevant_fields = {
+        'Today s Date': data.current_date.strftime("%B %#d, %Y"),
+        'Cheque Payable To print legibly': data.payable_to,
+        'In The Amount Of': "${0:.2f}".format(data.amount_cad),
+        'Describe the request andor provide additional information if necessary': data.request_desc,
+
+        'Requested By': data.request_by,
+        'Position': data.requester_position,
+    }
+
+    # contextual fields
+    if isinstance(data.receive_kind, Pickup):
+        relevant_fields['Picked up by'] = data.receive_kind.pickup_by
+        relevant_fields['Email'] = data.receive_kind.pickup_email
+    elif isinstance(data.receive_kind, Mailed):
+        relevant_fields['Street Address'] = data.receive_kind.street_address
+        relevant_fields['City, Province'] = data.receive_kind.city_province
+        relevant_fields['Postal Code'] = data.receive_kind.postal_code
+
+        # check boxes are not easy, so just update them in the reader's page
+        found_check_box_1 = False
+        found_check_box_3 = False
+        if "/Annots" in page:
+            for annotation in page["/Annots"]:
+                annotation_obj = annotation.get_object()
+                annotation_type = annotation_obj['/T']
+                if annotation_type == "Check Box1":
+                    found_check_box_1 = True
+                elif annotation_type == "Check Box3":
+                    found_check_box_3 = True
+
+                if annotation_type == "Check Box1" and data.receive_kind.mail_off_campus or \
+                   annotation_type == "Check Box3" and not data.receive_kind.mail_off_campus:
+                    annotation_obj.update({
+                        NameObject("/V"): NameObject("/Yes"),
+                        NameObject("/AS"): NameObject("/Yes")
+                    })
+        else:
+            raise KeyError("pdf has no checkboxes; something is wrong!")
+
+        if not found_check_box_1 or not found_check_box_3:
+            raise KeyError("could not find checkboxes; pdf may be malformed")
+    else:
+        raise AssertionError("Invalid value {}; exhaustive case".format(data.receive_kind))
+
+    # ------------------------
+    # validation
+
+    fields = reader.get_form_text_fields()
+    for key in relevant_fields.keys():
+        if key not in fields:
+            raise KeyError("pdf form did not have necessary key")
+
+    # ------------------------
+    # final writing
+
+    if not os.path.exists(tmp_dir):
+        os.makedirs(tmp_dir)
+
+    c = canvas.Canvas(tmp_dir + "watermark.pdf")
+    c.drawString(335, 702, "Computing Science Student Society")
+    if watermark_active:
+        c.drawImage(small_text_logo, 429, 725, 848 / 6, 174 / 6)
+    c.save()
+
+    # NOTE: what happens when multiple requests come in at once?
+    # I guess the file acts as a mutex & blocks one of the requests. This degraded performance is

```

Comment:
which file will act as a mutex and how would multiple requests come in? will this be used by more than just the treasurer?

Reply:
I might ask someone else to generate a cheque requisition for themselves, but the chance of two people using this tool at once is extremely small.

I should check what happens when `watermark.pdf` is read twice at once in `PdfReader(open(tmp_dir + "watermark.pdf", "rb"))`. I assumed that open() would pause the thread until the resource is open again, but it will probably actually fail. I'll look into it!

Reply:
I am uncomfortable with the idea of using `open` as a mutex. We'll have to think of something better.

Reply:
If it raises a "can't open file" error then we can handle it specifically, instead of just using it as a mutex, though I'd like to have it wait a few ms and try to access the file again at least

Reply:
hmm, that's sounds incredibly risky and I rather not have you do that. I'd rather you set a flag to let other requests know not to access the file. If you need help finding the best and safest way to accomplish this, let me know.